### ISSUE TITLE
Is there a reason why NLOPT_LD_CCSAQ is not included in nlotpr?

### ISSUE THREAD

**Issue by jchiquet** [astamm/nloptr#49]
Hi and thanks @jyypma for providing this great R interface to nlopt.

For some particular optimization problems, I use the CCSA quadratic variant of the method of moving asymptoc (MMA) implemented in nlopt. Indeed, I found the former to be slightly more efficient than the later on some problems. It is also more flexible as a user-defined Hessian can be supplied. See [the nlopt doc here about MMA and CCSAQ](https://nlopt.readthedocs.io/en/latest/NLopt_Algorithms/#mma-method-of-moving-asymptotes-and-ccsa).

Unfortunately, when I try to call this optimizer via nloptr, I get 

```r 
Error in is.nloptr(ret) : 
  Incorrect algorithm supplied. Use one of the following:
...
```

Any reason for not providing access to this optimizer in nloptr?

Here is a reproducible example

```r 
library('nloptr')

## Rosenbrock Banana function and gradient in separate functions
eval_f <- function(x) {
    return( 100 * (x[2] - x[1] * x[1])^2 + (1 - x[1])^2 )
}

eval_grad_f <- function(x) {
    return( c( -400 * x[1] * (x[2] - x[1] * x[1]) - 2 * (1 - x[1]),
                200 * (x[2] - x[1] * x[1])) )
}


# initial values
x0 <- c( -1.2, 1 )

opts <- list("algorithm"="NLOPT_LD_CCSAQ",
             "xtol_rel"=1.0e-8)
 
# solve Rosenbrock Banana function
res <- nloptr( x0=x0, 
               eval_f=eval_f, 
               eval_grad_f=eval_grad_f,
               opts=opts)
``` 

Thank for your time.

**Comment by jyypma** [comment]
This might be an algorithm that has been added at a later stage to NLopt.

Could you try running
```
# Hack to prevent checking on R side before sending to C.
R.utils::reassignInPackage("is.nloptr", pkgName="nloptr", function(x) { return(TRUE) })
```
before
```
# solve Rosenbrock Banana function
res <- nloptr( x0=x0, 
               eval_f=eval_f, 
               eval_grad_f=eval_grad_f,
               opts=opts)
```
and see what the output is?

On Windows, I get
```
Error: unknown algorithm NLOPT_LD_CCSAQ.
```
suggesting that the algorithm is not included in the NLopt version that is included in the Windows version of nloptr. This may be different on other operating systems.

**Comment by eddelbuettel** [comment]
Good point.  On Ubuntu with the distribution version 2.4.2 I see

```sh

edd@rob:~$ grep 'NLOPT_.._' /usr/include/nlopt.h
     NLOPT_GN_DIRECT = 0,
     NLOPT_GN_DIRECT_L,
     NLOPT_GN_DIRECT_L_RAND,
     NLOPT_GN_DIRECT_NOSCAL,
     NLOPT_GN_DIRECT_L_NOSCAL,
     NLOPT_GN_DIRECT_L_RAND_NOSCAL,
     NLOPT_GN_ORIG_DIRECT,
     NLOPT_GN_ORIG_DIRECT_L,
     NLOPT_GD_STOGO,
     NLOPT_GD_STOGO_RAND,
     NLOPT_LD_LBFGS_NOCEDAL,
     NLOPT_LD_LBFGS,
     NLOPT_LN_PRAXIS,
     NLOPT_LD_VAR1,
     NLOPT_LD_VAR2,
     NLOPT_LD_TNEWTON,
     NLOPT_LD_TNEWTON_RESTART,
     NLOPT_LD_TNEWTON_PRECOND,
     NLOPT_LD_TNEWTON_PRECOND_RESTART,
     NLOPT_GN_CRS2_LM,
     NLOPT_GN_MLSL,
     NLOPT_GD_MLSL,
     NLOPT_GN_MLSL_LDS,
     NLOPT_GD_MLSL_LDS,
     NLOPT_LD_MMA,
     NLOPT_LN_COBYLA,
     NLOPT_LN_NEWUOA,
     NLOPT_LN_NEWUOA_BOUND,
     NLOPT_LN_NELDERMEAD,
     NLOPT_LN_SBPLX,
     NLOPT_LN_AUGLAG,
     NLOPT_LD_AUGLAG,
     NLOPT_LN_AUGLAG_EQ,
     NLOPT_LD_AUGLAG_EQ,
     NLOPT_LN_BOBYQA,
     NLOPT_GN_ISRES,
     NLOPT_LD_SLSQP,
     NLOPT_LD_CCSAQ,
     NLOPT_GN_ESCH,
edd@rob:~$
```

**Comment by jchiquet** [comment]
@jyypma 

> Could you try running
> 
> ```
> # Hack to prevent checking on R side before sending to C.
> R.utils::reassignInPackage("is.nloptr", pkgName="nloptr", function(x) { return(TRUE) })
> ```
> before
> 
> ```
> # solve Rosenbrock Banana function
> res <- nloptr( x0=x0, 
>                eval_f=eval_f, 
>                eval_grad_f=eval_grad_f,
>                opts=opts)
> ```
> and see what the output is?
> 
> On Windows, I get
> 
> ```
> Error: unknown algorithm NLOPT_LD_CCSAQ.
> ```
> suggesting that the algorithm is not included in the NLopt version that is included in the Windows version of nloptr. This may be different on other operating systems.

Got the same error on Ubuntu 18.04 : I also had to modify src/nlopt.c to make things work. I sent you a PR #52 if you want to consider my solution.

**Comment by jyypma** [comment]
Closed by 507c2625c320f5bd7d66c37fce1b16a72e2b57ac
